activemq-client/src/main/java/org/apache/activemq/command/ActiveMQBytesMessage.java

```diff
@@ -956,4 +956,22 @@ protected void doCompress() throws IOException {
             }
         }
     }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean isBodyAssignableTo(Class c) {
+        return getContent() == null || c.isAssignableFrom(byte[].class);
+    }
+
+    @SuppressWarnings("unchecked")
+    protected <T> T doGetBody(Class<T> asType) {
+        //Make sure the bytes are stored before trying to copy and return
+        if (dataOut != null && getContent() == null) {
+            storeContent();
+        }
+
+        final ByteSequence content = getContent();
+        return content != null ? (T) new ByteSequence(content.getData(), content.getOffset(),
+            content.getLength()).getData() : null;
```

Comment:
This isn't a deep copy from what I see in the ByteSequence code, just a shallow wrapping of the array, plus it creates a new ByteSequence that would never actually be used so needless GC.  I'd just do an Arrays.copyOfRange on the data in the original since you want t deep copy here to avoid the caller modifying the bytes in the message content as you want to get the same payload on a rollback / recover etc.

Reply:
Thanks, good catch. I just copied this from the parent Message class in its copy() method:

https://github.com/apache/activemq/blob/ae4e305f850dd35d92f8e49758309f21609d3206/activemq-client/src/main/java/org/apache/activemq/command/Message.java#L184-L189

This makes me wonder if this needs to be fixed there at some point as well. I've sometimes seen weird behavior with the VM transport that appeared to be from shared messages between server/client even though copying should be happen so maybe this is a culprit.

Reply:
(also if there is somehow any offset/length diff then the returned bytes would actually be invalid as they dont account for those)

Reply:
It's going to depend on how that is really used, a shallow copy may be sufficient for the JMS 1.1 APIs where you didn't get direct access as you are getting in this simplified API method where you can accidentally hand out the byte[] payload reference.

Reply:
Is using the readBytes() an option here?

```
    @Override
    public <T> T getBody(Class<T> c) throws JMSException {
        if(c == null || getContent() == null) { return null; }

        byte[] tmpBytes = new byte[(int)getBodyLength()];
        readBytes(tmpBytes);
        return (T)tmpBytes;
    }
```

Reply:
I'm going to add a copy method to ByteSequence, that seems simplest.